### PR TITLE
fix(tree): doc corruption when rebasing over compositions of MoveIn&MoveOut marks

### PR DESCRIPTION
**Pull request by yann-achard-MS**
## Description

This PR introduces a new kind of mark ("Rename") to represent the renaming of an empty cell.
Rename is used instead of AttachAndDetach marks that represent a MoveIn and MoveOut.

This PR also fixes a bug in the composition of such marks with further moves (see https://dev.azure.com/fluidframework/internal/_workitems/edit/8209)

Rename could also be used to represent the impact of other kinds of AttachAndDetach pairings. This will be tackled in later PRs.

This PR introduces a new format version (V3) for sequence changeset, which rolls up to a new top-level format version for SharedTree (V4). This top-level V4 version is not yet offered as an option for writing (it is only available for reading). There are two motivations for this:
1. The longer we keep a configuration where ST can read V4 before actually writing V4, the less likely it is that users will run into situations where cross-version collab breaks older clients.
2. We may want to make further changes to the V4 version before publishing it (thought that diminishes the value of motivation 1 when we do).

## Breaking Changes

None:
* Forwards Compat: the old formats have not been altered and the new formats are not yet writable, so old clients should not fail when dealing with new clients or data generated by new clients.
* Backwards Compat: the decoders for old formats have been updated to normalize these formats into the new format, so new clients should not fail when dealing with old clients or data generated by old clients.

## Reviewer Guidance

fomatV3.ts and sequenceFieldCodecV3.ts are clones of formatV2.ts and sequenceFieldCodecV2.ts prior to this PR with the addition of the support for Rename marks.

### CODE REVIEW

File: packages/dds/tree/src/feature-libraries/sequence-field/compose.ts [hunk 1]
```diff
@@ -324,11 +347,6 @@ function composeMarksIgnoreChild(
 	}
 
 	if (!markHasCellEffect(baseMark) && !markHasCellEffect(newMark)) {
-		if (isNoopMark(baseMark)) {
```
Comment (alex-pardes) [inline packages/dds/tree/src/feature-libraries/sequence-field/compose.ts:327]: What was the motivation for moving this code instead of putting the rename logic in here?

Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/compose.ts:327]: The block of logic for handling rename marks gets to be more cohesive (as opposed to adding logic inside multiple existing branches) if it appears earlier. It also gets to be simpler if the noop marks are handled before it. Since handling the noop mark is itself simple and self-contained it seemed reasonable to order things that way.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts [hunk 2]
```diff
@@ -239,6 +240,8 @@ class RebaseQueue {
 function addMovedMarkEffect(mark: Mark, effect: MarkEffect): Mark {
 	if (isMoveIn(mark) && isMoveOut(effect)) {
 		return { ...mark, type: "Insert" };
+	} else if (isRename(mark) && isMoveOut(effect)) {
+		return { ...effect, count: mark.count, idOverride: mark.idOverride };
```
Comment (alex-pardes) [inline packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:244]: We can do this in a separate PR, but we can probably adjust the type for moved marks so that it's clear that `effect` is a type that extends `DetachFields`.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts [hunk 3]
```diff
@@ -352,11 +353,19 @@ function separateEffectsForMove(mark: MarkEffect): {
...
 	switch (type) {
 		case "Remove":
-		case "MoveOut":
+		case "MoveOut": {
+			if (mark.idOverride !== undefined) {
+				const remains: MarkEffect = { type: "Rename", idOverride: mark.idOverride };
+				const follows: Mutable<MarkEffect> = { ...mark };
+				delete follows.idOverride;
+				return { remains, follows };
+			}
```
Comment (alex-pardes) [inline packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:362]: I think it would be nice to have a comment explaining why we have to do this.

File: packages/dds/tree/src/feature-libraries/sequence-field/compose.ts [hunk 4]
```diff
@@ -179,8 +180,30 @@ function composeMarksIgnoreChild(
...
+		return newMark;
+	} else if (isNoopMark(newMark)) {
+		return baseMark;
+	}
+
+	if (isRename(baseMark) && isRename(newMark)) {
+		return { ...baseMark, idOverride: newMark.idOverride };
+	} else if (isRename(baseMark)) {
+		assert(isAttach(newMark) || isAttachAndDetachEffect(newMark), "Unexpected mark type");
+		return { ...newMark, cellId: baseMark.cellId };
```
Comment (alex-pardes) [inline packages/dds/tree/src/feature-libraries/sequence-field/compose.ts:193]: Are rename marks always supposed to have a cell ID? Do we assume that rename marks are never rebased?

File: packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldCodecV1.ts [hunk 5]
```diff
@@ -130,6 +132,16 @@ export function makeV1Codec(
 							detach: markEffectCodec.encode(effect.detach, context) as Encoded.Detach,
 						},
 					};
+				case "Rename":
+					// In documents post-2024-07-23, renames are encoded as AttachAndDetach with a special id.
```
Comment (alex-pardes) [inline packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldCodecV1.ts:136]: Do you mean documents before this date? What is the significance of this date?

File: packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldCodecV1.ts [hunk 6]
```diff
@@ -285,13 +307,29 @@ export function makeV1Codec(
 				if (mark.cellId !== undefined) {
 					decodedMark.cellId = cellIdCodec.decode(mark.cellId, context.baseContext);
 				}
+				// This assignment is solely to recover the type of `decodedMark`.
```
Comment (alex-pardes) [inline packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldCodecV1.ts:310]: I don't understand what this means.

File: packages/dds/tree/src/feature-libraries/sequence-field/compose.ts (hunk 4 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/compose.ts:193]: They are (same as Attach&Detach marks).

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts (hunk 2 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:244]: Created https://dev.azure.com/fluidframework/internal/_workitems/edit/9387

File: packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldCodecV1.ts (hunk 5 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldCodecV1.ts:136]: It's an approximation for the date this PR goes in.

Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldCodecV1.ts:136]: Wouldn't it make more sense to talk in terms of releases instead of dates? Or just in terms codec versions?

**Review by alex-pardes — APPROVED** [review]

File: packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldCodecV1.ts (hunk 5 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldCodecV1.ts:136]: I like the suggestion of talking in terms of release so I'll update the comments to do that.
What do you mean by codec version? If you mean "one of V1/V2/V3" (which is how the term is used in the codebase) then that's not what comment is about. Instead it is about the different implementations of the same (in this case V1) codec over time (before and after this PR).

File: packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldCodecV1.ts (hunk 6 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/sequenceFieldCodecV1.ts:310]: The type system thinks that the type of `decodedMark` is `NoopMark & HasMarkFields`. This prevents the if block below from compiling.
To circumvent that, I am casting `decodedMark` back to a `Mark`.
I was concerned that such a assignment would look weird to anyone reading the code. They might either think that it is unecessary (as it should be) or that it is needed for some very fishy reason that is hard to see.
Let me know if you have suggestions on how to get the message acrross more clearly.

**Review by alexvy86 — APPROVED** [review]
Approving for cr-docs. Only looked at the changeset.

If we have more details about the scenario that was fixed which could be useful to people reading the release notes, I think they should be added to the changeset (e.g. "if you were doing X, before this change the document would get corrupted, but now it won't"). But the change summary is good for now.
